core/services.go

```diff
@@ -303,18 +312,20 @@ func (ss *Services) Stop(ctx context.Context, bk *buildkit.Client, svc *Service)
 // It is called when a client is closing.
 func (ss *Services) StopClientServices(ctx context.Context, client *engine.ClientMetadata) error {
 	ss.l.Lock()
-	defer ss.l.Unlock()
-
-	eg := new(errgroup.Group)
+	var svcs []*RunningService
 	for _, svc := range ss.running {
-		if svc.Key.ClientID != client.ClientID {
-			continue
+		if svc.Key.ClientID == client.ClientID {
+			svcs = append(svcs, svc)
 		}
+	}
+	ss.l.Unlock()
 
+	eg := new(errgroup.Group)
+	for _, svc := range svcs {
 		svc := svc
 		eg.Go(func() error {
 			bklog.G(ctx).Debugf("shutting down service %s", svc.Host)
-			if err := svc.Stop(ctx); err != nil {
+			if err := ss.stopGraceful(ctx, svc, DetachGracePeriod); err != nil {
```

Comment:
Debatable, but I probably would have gone with force killing the services in this codepath (which should only matter when the client has disconnected but left some services running). If the client didn't care enough to actually stop their services then they probably just want to shutdown asap rather than wait for anything.

Reply:
Ahhh this is definitely debatable, but I think it makes sense to avoid blocking close as much as possible.

We *could* gracefully shutdown in the background... but now any stopping messages/logs get lost, without the client being able to see them, so I don't love that.

I think SIGKILL is the right call here - users should stop their own services if they really care about this case.

---

So in that case, we'd only apply the graceful stop when detaching - when all the service bindings are let go. But - does that even make sense? A common case will be that the client immediately disconnects after a run of a container that had service bindings, so we'd end up skipping that graceful exit, and apply the force-kill here.

Maybe the only way to get `SIGTERM` is to manually call `Stop` - if users don't do this, we always `SIGKILL`. `stopGraceful` isn't then necessary, which is somewhat nice.